### PR TITLE
JAMES-2864 MockSMTPServer POJOs

### PR DESCRIPTION
**Pull request by trantienduchn**
Introducing new POJOs which are used to setup mock smtp server behaviors....

### CODE REVIEW

**Comment by trantienduchn** [comment]
```
[cf01797a5db833639130d0ea312d0a8132632720] [ERROR] Errors: 
[cf01797a5db833639130d0ea312d0a8132632720] [ERROR]   ObjectStorageBlobsDAOTest.saveConcurrentlyWithNonPreExistingBucketShouldNotFail » NotTerminated
```
test this please

**Comment by chibenwa** [comment]
```
ObjectStorageBlobsDAOTest.saveConcurrentlyWithNonPreExistingBucketShouldNotFail » NotTerminated
```

test this please

**Comment by chibenwa** [comment]
test this please

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/Operator.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mock.smtp.server;
+
+public interface Operator {
+    Operator CONTAINS = String::contains;
+
+    boolean matches(String testedValue, String referenceValue);
```
Comment (mbaechler) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/Operator.java]: could we use the usual wording: actual/expected ?

Comment (mbaechler) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/Operator.java]: what about having a fluent API?

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/Response.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mock.smtp.server;
+
+import com.google.common.base.Preconditions;
+
+class Response {
+    static  class SMTPStatusCode {
```
Comment (mbaechler) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/Response.java]: would it make sense to have the status as an enum of all known valid statuses?

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPBehavior.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+package org.apache.james.mock.smtp.server;
+
+import java.util.Optional;
+
+import com.google.common.base.Preconditions;
+
+public class MockSMTPBehavior {
+    public static final class NumberOfAnswersPolicy {
+
+        public static NumberOfAnswersPolicy always() {
```
Comment (mbaechler) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPBehavior.java]: anytime?

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPBehavior.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        this.condition = condition;
+        this.response = response;
+        this.numberOfAnswers = numberOfAnswers;
+    }
+
+    public SMTPCommand getSmtpCommand() {
+        return smtpCommand;
+    }
+
+    public Optional<Condition> getCondition() {
```
Comment (mbaechler) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPBehavior.java]: I think that "no condition" could be encoded as a condition

Comment (trantienduchn) [reply server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPBehavior.java]: "no condition" make me really think there is no condition. I would name it matchAll condition. but they are acting the same.

**Comment by chibenwa** [comment]
```
Progress (5): 2.2/4.4 MB | 2.0/2.9 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project metrics-api: Fatal error compiling: invalid flag: --release -> [Help 1]
```

Rebase needed?

**Comment by trantienduchn** [comment]
```
[7ff0d4e5d93b058fd6060481bedc174a5cf3267d] [ERROR] Errors: 
[7ff0d4e5d93b058fd6060481bedc174a5cf3267d] [ERROR]   MemoryUploadCucumberTest » Socket Broken pipe (Write failed)
```
this test fails too many times, test this please

**Comment by chibenwa** [comment]
`MemoryUploadCucumberTest » Socket Broken pipe (Write failed)` test this please

**Comment by trantienduchn** [comment]
```
[7ff0d4e5d93b058fd6060481bedc174a5cf3267d] [ERROR] Errors: 
[7ff0d4e5d93b058fd6060481bedc174a5cf3267d] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:495 » Execution
[7ff0d4e5d93b058fd6060481bedc174a5cf3267d] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:495 » Execution
```
test this please

**Comment by trantienduchn** [comment]
```
[7ff0d4e5d93b058fd6060481bedc174a5cf3267d] [ERROR] Errors: 
[7ff0d4e5d93b058fd6060481bedc174a5cf3267d] [ERROR]   MemoryUploadCucumberTest » Socket Broken pipe (Write failed)
```
test this please

**Comment by trantienduchn** [comment]
again test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**